### ISSUE TITLE
Deep code search

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#388]
Deep code search<br>
<br>
https://ift.tt/2Kph26w<br>
<br>
adriancolyer<br>
<br>
<div>
<div class="entry clear">
<p><a href="https://guxd.github.io/papers/deepcs.pdf">Deep code search</a> Gu et al., <em>ICSE&rsquo;18</em></p>
<p>The problem with searching for code is that the query, e.g. &ldquo;read an object from xml,&rdquo; doesn&rsquo;t look very much like the source code snippets that are the intended results, e.g.:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-example-snippet.jpeg?w=480">*</p>
<p>That&rsquo;s why we have Stack Overflow! Stack Overflow can help with &lsquo;how to&rsquo; style queries, but it can&rsquo;t help with searches inside codebases you care about. For example, &ldquo;where in this codebase are events queued on a thread?&rdquo;</p>
<blockquote>
<p>&hellip;an effective code search engine should be able to understand the semantic meanings of natural language queries and source code in order to improve the accuracy of code search.</p>
</blockquote>
<p><em>DeepCS</em> is just such a search engine for code, based on the <em>CODEnn</em> (Code-Description Embedding Neural Network) network model. During training, it takes code snippets (methods) and corresponding natural language descriptions (from the method comments) and learns a <em>joint-embedding</em>. I.e., it learns embeddings such that a method description and its corresponding code snippet are both mapped to a similar point in the same shared embedding space. Then given a natural language query, it can embed the query in vector space and look for nearby code snippets. Compared to Lucene powered code search tools, and the recently proposed state-of-the-art CodeHow search tool, CODEnn gives excellent results.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-8.jpeg?w=480"></p>
<h3>Embeddings</h3>
<p>One of the most popular posts over time on this blog has been &lsquo;<a href="https://blog.acolyer.org/2016/04/21/the-amazing-power-of-word-vectors/">The amazing power of word vectors</a>,&rsquo; which describes the process of turning words into vectors. To turn a <em>sequence</em> of words into a vector, one common approach is to use an RNN (e.g., LSTM). For example, given a sentence such as &lsquo;parse xml file,&rsquo; the RNN reads the first word, &lsquo;parse,&rsquo; maps it into a vector <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7Bw%7D_1&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\mathbf{w}_1" title="\mathbf{w}_1" class="latex">, and then computes the RNN hidden state <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7Bh%7D_1&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\mathbf{h}_1" title="\mathbf{h}_1" class="latex"> using <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7Bw%7D_1&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\mathbf{w}_1" title="\mathbf{w}_1" class="latex">. Then it reads the second word, &lsquo;xml,&rsquo; maps that into word vector <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7Bw%7D_2&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\mathbf{w}_2" title="\mathbf{w}_2" class="latex"> and updates the hidden state <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7Bh%7D_1&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\mathbf{h}_1" title="\mathbf{h}_1" class="latex"> to <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7Bh%7D_2&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\mathbf{h}_2" title="\mathbf{h}_2" class="latex">. We k keep going in this manner until we reach the end of the sentence, and then use the final hidden state ($\mathbf{h}_3$ in this example) as the sentence embedding.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-1.jpeg?w=520"></p>
<p>We don&rsquo;t want just any old embedding though. We want to learn embeddings such that code snippets (think of them like &lsquo;code sentences&rsquo; for now, we&rsquo;ll get to the details shortly) and their corresponding descriptions have nearby embeddings.</p>
<blockquote>
<p>Joint Embedding, also known as multi-model embedding, is a technique to jointly embed/correlate heterogeneous data into a unified vector space so that semantically similar concepts across the two modalities occupy nearby regions of the space.</p>
</blockquote>
<p>When we&rsquo;re training, we use a similarity measure (e.g. cosine) in the loss function to encourage the joint mapping.</p>
<h3>CODEnn high level architecture</h3>
<p>CODEnn has three main components: a code embedding network to embed source code snippets in vectors; a description embedding network to embed natural language descriptions into vectors; and a similarity module (cosine) to measure the degree of similarity between code and descriptions. At a high level it looks like this:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-4a.jpeg?w=400"></p>
<p>If we zoom in one level, we can start to see the details of the network constructions:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-4b.jpeg?w=640"></p>
<h3>The code embedding network</h3>
<p>A code snippet (a method) is turned into a tuple <em>(M,A,T)</em> where <em>M</em> is a sequence of camelCase split tokens in the method name, <em>A</em> is an <em>API sequence</em> (the method invocations made by the method body), and <em>T</em> is the set of tokens in the snippet.</p>
<p>The method name and API invocation sequences are both embedded (separately) into vectors using RNNs with maxpooling. The tokens have no strict order, so they are embedded using a conventional multilayer perceptron. The three resulting vectors are then fused into one vector through a fully connected layer.</p>
<h3>The description embedding network</h3>
<p>The description embedding network uses an RNN with maxpooling to encode the natural language description from the (first sentence of) the method comment.</p>
<h3>Training</h3>
<p>During training, training instances are given as triples (C, D+, D-), where C is a code snipped, D+ is the correct (actual) description of C, and D- is an incorrect description, chosen randomly from the pool of all descriptions. The loss function seeks to maximise the cosine similarity between C and D+, and make the cosine similarity between C and D- as large as possible.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-5.jpeg?w=640"></p>
<p>The training corpus is based on open-source Java projects on GitHub, resulting in a body of just over 18M commented Java methods.</p>
<blockquote>
<p>For each Java method, we use the method declaration as the code element and the first sentence of its documentation comment as its natural language description. According to the Javadoc guidance, the first sentence is usually a summary of a method.</p>
</blockquote>
<p>The method bodies are parsed using the Eclipse JDT compiler. (Note, some of the tools from <a href="https://sourced.tech/projects/#retrieval">Source{d}</a> may also be useful here).</p>
<p>All the RNNs are embodied as LSTMs with 200 hidden units in each direction, and word embeddings have 100 dimensions.</p>
<h3>Indexing and querying</h3>
<p>To index a codebase, DeepCS embeds all code snippets in the codebase into vectors using offline processing. Then during online searching DeepCS embeds the natural language user query, and estimates the cosine similarities between the query embedding and pre-computed code snippet embeddings. The top K closest code snippets are returned as the query results.</p>
<h3>Evaluation</h3>
<p>Evaluation is done using a search codebase constructed from ~10 thousand Java projects on GitHub (different to the training set), with at least 20 stars each. All code is indexed (including methods without any comments), resulting in just over 16M indexed methods.</p>
<blockquote>
<p>We build a benchmark of queries from the top 50 voted Java programming questions in Stack Overflow. To achieve so, we browse the list of Java-tagged questions in Stack Overflow and sort them according to the votes that each one receives.</p>
</blockquote>
<p>To qualify, the questions must be about a concrete Java programming task and include a code snippet in the accepted answer. The 50 such resulting questions are shown in the following table. In the right-hand column we can see the <em>FRank</em> (rank of the first hit result in the result list) for DeepCS, CodeHow, and a conventional Lucene-based search tool.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-table-1.jpeg?w=640"></p>
<p>Here are some representative query results.</p>
<ul>
<li>&ldquo;<em>queue an event to be run on the thread</em>&rdquo; vs &ldquo;<em>run an event on a thread queue</em>&rdquo; (both have similar words in the query, but are quite different).</li>
</ul>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-9.jpeg?w=520"></p>
<ul>
<li>&ldquo;<em>get the content of an input stream as a string using a specified character encoding</em>&rdquo;</li>
</ul>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-10.jpeg?w=520"></p>
<ul>
<li>&ldquo;<em>read an object from an xml file</em>&rdquo; (note that the words xml, object, and read don&rsquo;t appear in the result, but DeepCS still finds it)</li>
</ul>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-11a.jpeg?w=520"></p>
<ul>
<li>&ldquo;<em>play a song</em>&rdquo; (another example of associative search, where DeepCS can recommend results with semantically related words such as <em>audio</em>).</li>
</ul>
<p><img src="https://adriancolyer.files.wordpress.com/2018/06/deepcs-fig-11b.jpeg?w=520"></p>
<p>DeepCS isn&rsquo;t perfect of course, and sometimes ranks partially relevant results higher than exact matching ones.</p>
<blockquote>
<p>This is because DeepCS ranks results by just considering their semantic vectors. In future work, more code features (such as programming context) could be considered in our model to further adjust the results.</p>
</blockquote>
</div>
</div>
<br>
<br>
<br>
<br>
BigData,AI|ML|DM|IR|PR<br>
<br>
via the morning paper https://ift.tt/1WJ1E4W<br>
<br>
June 26, 2018 at 01:00PM
